From 593f1a79527c54773d641318c27a26830e90371b Mon Sep 17 00:00:00 2001
From: Harsh J <harsh@cloudera.com>
Date: Tue, 25 Aug 2015 11:21:19 +0530
Subject: [PATCH 1202/2848] HADOOP-12352. Delay in checkpointing Trash can
 leave trash for 2 intervals before deleting.
 Contributed by Casey Brotherton.

Change-Id: Iae66e83f6194700afb1b57c4d559e103a3404cbc
---
 .../org/apache/hadoop/fs/TrashPolicyDefault.java   |   11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/TrashPolicyDefault.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/TrashPolicyDefault.java
index d6a9b4b..efe6efd 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/TrashPolicyDefault.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/TrashPolicyDefault.java
@@ -164,12 +164,19 @@ public boolean moveToTrash(Path path) throws IOException {
   @SuppressWarnings("deprecation")
   @Override
   public void createCheckpoint() throws IOException {
+    createCheckpoint(new Date());
+  }
+
+  @SuppressWarnings("deprecation")
+  public void createCheckpoint(Date date) throws IOException {
+
     if (!fs.exists(current))                     // no trash, no checkpoint
       return;
 
     Path checkpointBase;
     synchronized (CHECKPOINT) {
-      checkpointBase = new Path(trash, CHECKPOINT.format(new Date()));
+      checkpointBase = new Path(trash, CHECKPOINT.format(date));
+
     }
     Path checkpoint = checkpointBase;
 
@@ -286,7 +293,7 @@ public void run() {
                 TrashPolicyDefault trash = new TrashPolicyDefault(
                     fs, home.getPath(), conf);
                 trash.deleteCheckpoint();
-                trash.createCheckpoint();
+                trash.createCheckpoint(new Date(now));
               } catch (IOException e) {
                 LOG.warn("Trash caught: "+e+". Skipping "+home.getPath()+".");
               } 
-- 
1.7.9.5

